### DISCUSSION TITLE
Setting x-text values to AlpineJS' $store varables before AlpineJS has loaded

### DISCUSSION THREAD

**Discussion by anjanesh** [alpinejs/alpine#3524] [1. Help] [answered]
How do I do this when AlpineJS is deferred - not fully loaded yet.

```html
<li x-text="$store.global_data.x']"></li>
```

I have already done :

```javascript
document.addEventListener('alpine:init', () =>
{
    Alpine.store('global_data',
    {
        data: { x: '', y: '', z: '' },
        foo()
        {
        },
    });
});
```

I get

**Alpine Expression Error: Cannot read properties of undefined (reading 'x')**

**Comment by ekwoka** [comment]
Is that the only error?

This shouldn't be happening unless you're not hooking up this listener before Alpine has initialized.

So the order things are happening is incorrect and that can't be seen properly from these snippets

**Reply by anjanesh** [reply]
Its working now - I had a different issue though - even for setting global variables, `$store.global_data.x`, the DOM element has to be within and element having `x-data` too.
```
<div x-data="{}">
<li x-text="$store.global_data.x"></li>
</div>
```

**Reply by ekwoka** [reply, marked as answer]
Yes, all directives (except for `x-init`) need to be within a component context.

You can make empty components with just `x-data` as well, don't need the `="{}"` fyi

requiring the alpine context just helps reduce the amount of tree walking Alpine needs to do. (you can of course slap an `x-data` on the body to "opt out" of this optimization)
